crates/matrix-sdk/src/encryption/mod.rs

```diff
@@ -404,10 +404,13 @@ impl Client {
     fn get_dm_room(&self, user_id: &UserId) -> Option<room::Joined> {
         let rooms = self.joined_rooms();
         let room_pairs: Vec<_> =
-            rooms.iter().map(|r| (r.room_id().to_owned(), r.direct_target())).collect();
+            rooms.iter().map(|r| (r.room_id().to_owned(), r.direct_targets())).collect();
         trace!(rooms = ?room_pairs, "Finding direct room");
 
-        let room = rooms.into_iter().find(|r| r.direct_target().as_deref() == Some(user_id));
+        // Find the room we share with the `user_id` and only with `user_id`
+        let room = rooms
+            .into_iter()
+            .find(|r| r.direct_targets().len() == 1 && r.direct_targets().contains(user_id));
```

Comment:
Creating the `direct_targets()` twice is a bit heavy (as they create a copy every time). Let's do that only once:
```suggestion
            .find(|r| {
                 let targets = r.direct_targets();
                 targets.len() == 1 && targets.contains(user_id)
            });
```